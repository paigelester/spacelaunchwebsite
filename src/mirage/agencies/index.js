import AgencySeed from './seed';
import AgencyModels from './models';
import AgencyRoutes from './routes';
import AgencyFactories from './factories';

export default {
    seed: AgencySeed,
    models: AgencyModels,
    routes: AgencyRoutes,
    factories: AgencyFactories
}