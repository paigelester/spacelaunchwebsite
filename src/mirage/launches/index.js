import LaunchSeed from './seed';
import LaunchModels from './models';
import LaunchRoutes from './routes';
import LaunchFactories from './factories';

export default {
    seed: LaunchSeed,
    models: LaunchModels,
    routes: LaunchRoutes,
    factories: LaunchFactories
}