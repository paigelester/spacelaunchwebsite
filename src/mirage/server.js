import { createServer, Model, Factory, trait, Response } from 'miragejs';

export const makeServer = () => {
    let server = createServer({

        models: {
            launch: Model
        },

        factories: {
            launch: Factory.extend({
                name(i) { return `Launch name ${i + 1}`; },
                image: 'https://spacelaunchnow-prod-east.nyc3.digitaloceanspaces.com/media/launch_images/starship2520sn_image_20210107083733.jpeg',
                status: {
                    name(i) {
                        let statuses = ['Go for launch', 'To be confirmed'];

                        return statuses[i % statuses.length];
                    },
                },
                window_start(i) {
                    const startDate = new Date();
                    startDate.setDate(startDate.getDate() + (i + 1));

                    return startDate;
                },
                window_end(i) {
                    const endDate = new Date();
                    endDate.setDate(endDate.getDate() + (i + 2));

                    return endDate;
                },
                launch_service_provider: {
                    name: 'SpaceX',
                    type: 'Commercial',
                    url: 'https://ll.thespacedevs.com/2.1.0/agencies/121/'
                },

                previous: trait({
                    window_start(i) {
                        const startDate = new Date();
                        startDate.setDate(startDate.getDate() - (i + 2));

                        return startDate;
                    },
                    window_end(i) {
                        const endDate = new Date();
                        endDate.setDate(endDate.getDate() - (i + 1));

                        return endDate;
                    }
                })
            })
        },

        seeds(dataServer) {
            dataServer.createList('launch', 10);
            dataServer.createList('launch', 10, 'previous');
        },

        routes() {
            this.urlPrefix = 'https://ll.thespacedevs.com/2.1.0/'

            this.get('launch/:id', (schema, request) => {
                const launchId = request.params.id;

                const launch = schema.launches.find(launchId);
                
                if (launch === null) {
                    return Response(404);
                }

                return launch.attrs;
            });

            this.get("launch/upcoming", (schema, request) => {
                const limit = request.queryParams.limit;

                const upcomingLaunches = schema.launches.all().models.reduce((prev, current) => {
                    if (current.window_start > new Date()) {
                        prev.push(current);
                    }

                    return prev;
                }, []).slice(0, limit);

                return {
                    count: upcomingLaunches.length,
                    results: upcomingLaunches
                };
            });
        }

    });

    return server;
};