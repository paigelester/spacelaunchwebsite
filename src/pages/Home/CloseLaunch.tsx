import React from "react";
import { Card } from "react-bootstrap";

import LaunchResource from 'api/LaunchResource';

import LaunchStatus from "components/features/LaunchStatusIndicator";

export enum CloseLaunchType {
    Upcoming = "Upcoming",
    Previous = "Previous"
}

interface CloseLaunchProps {
    type: CloseLaunchType;
    launch: LaunchResource;
}

const CloseLaunch = (props: CloseLaunchProps) => {
    return (
        <Card>
            <Card.Header>
                <div className="p-2 pb-3 border-bottom border-light">
                    <h5 className="d-inline">{props.type} launch</h5>
                    {props.launch.status && <div className="float-end"><LaunchStatus status={props.launch.status} /></div>}
                </div>
            </Card.Header>
            <Card.Body>
                <h5>{props.launch.name}</h5>
                <p>{props.launch.mission?.description}</p>
            </Card.Body>
        </Card>
    );
};

export default CloseLaunch;