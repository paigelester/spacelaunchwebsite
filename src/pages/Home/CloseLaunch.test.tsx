import React from "react";
import { render } from "@testing-library/react";
import { makeServer } from 'mirage/server';
import CloseLaunch, { CloseLaunchType } from "./CloseLaunch";

let server: any;
beforeEach(() => { server = makeServer(); });
afterEach(() => { server.shutdown() });

it("renders as expected", () => {
    // given
    const launch = server.create('launch');

    // when
    const { container } = render(
        <CloseLaunch launch={launch} type={CloseLaunchType.Upcoming} />
    );

    // then
    expect(container).toMatchSnapshot();
});

it("renders title as expected given close launch is an upcoming launch", () => {
    // given
    const launch = server.create('launch');

    // when
    const { getByText } = render(
        <CloseLaunch launch={launch} type={CloseLaunchType.Upcoming} />
    );

    // then
    expect(getByText('Upcoming launch')).toBeTruthy();
});

it("renders title as expected given close launch is a previous launch", () => {
    // given
    const launch = server.create('launch');

    // when
    const { getByText } = render(
        <CloseLaunch launch={launch} type={CloseLaunchType.Previous} />
    );

    // then
    expect(getByText('Previous launch')).toBeTruthy();
});
