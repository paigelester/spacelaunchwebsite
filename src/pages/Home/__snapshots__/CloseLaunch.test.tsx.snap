// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="p-2 pb-3 border-bottom border-light"
      >
        <h5
          class="d-inline"
        >
          Upcoming
           launch
        </h5>
        <div
          class="float-end"
        >
          <span
            class="badge bg-info"
          >
            To Be Determined
          </span>
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <h5>
        Falcon 9 Block 5 | Starlink 21
      </h5>
      <p>
        First cargo delivery mission to the space station.
      </p>
    </div>
  </div>
</div>
`;
